lifetimes/utils.py

```diff
@@ -101,31 +98,93 @@ def summary_data_from_transaction_data(transactions, customer_id_col, datetime_c
+        first_transactions[[customer_id_col, datetime_col, 'first']],
+        how="left",
+        on=[customer_id_col, datetime_col]
+    )
+    # boolean invert the 'first' column to give a column representing repeated transactions
+    period_transactions['repeated'] = ~(period_transactions['first'].fillna(False))
+    select_columns.append('repeated')
 
-    # count all orders by customer.
-    customers = period_transactions[datetime_col].groupby(level=customer_id_col, sort=False).agg(['max', 'min', 'count'])
+    return period_transactions[select_columns]
 
-    # subtract 1 from count, as we ignore their first order.
-    customers['frequency'] = customers['count'] - 1
 
+def summary_data_from_transaction_data(transactions, customer_id_col, datetime_col, monetary_value_col=None, datetime_format=None,
+                                       observation_period_end=datetime.today(), freq='D'):
+    """
+    This transforms a Dataframe of transaction data of the form:
+        customer_id, datetime [, monetary_value]
+    to a Dataframe of the form:
+        customer_id, frequency, recency, T [, monetary_value]
+    Parameters:
+        transactions: a Pandas DataFrame.
+        customer_id_col: the column in transactions that denotes the customer_id
+        datetime_col: the column in transactions that denotes the datetime the purchase was made.
+        monetary_value_col: the columns in the transactions that denotes the monetary value of the transaction.
+            Optional, only needed for customer lifetime value estimation models.
+        observation_period_end: a string or datetime to denote the final date of the study. Events
+            after this date are truncated.
+        datetime_format: a string that represents the timestamp format. Useful if Pandas can't understand
+            the provided format.
+        freq: Default 'D' for days, 'W' for weeks, 'M' for months... etc. Full list here:
+            http://pandas.pydata.org/pandas-docs/stable/timeseries.html#dateoffset-objects
+    """
+    observation_period_end = pd.to_datetime(observation_period_end, format=datetime_format).to_period(freq)
+
+    # label all of the repeated transactions
+    repeated_transactions = find_repeated_transactions(
+        transactions,
+        customer_id_col,
+        datetime_col,
+        monetary_value_col,
+        datetime_format,
+        observation_period_end,
+        freq
+    )
+    # group together all transactions by customer_id and find each customers first and last transaction
+    customers = repeated_transactions.groupby(customer_id_col)[datetime_col].agg(['min', 'max'])
+    # count up all the repeated transactions for each customer
+    # by using a pivot_table, customers with zero repeated transactions are naturally accounted for
+    customers['frequency'] = pd.pivot_table(
+        repeated_transactions,
+        index=customer_id_col,
+        values='repeated',
+        aggfunc=np.sum
+    )
```

Comment:
I've changed the frequency calculation back to `customers['frequency'] = customers['count'] - 1` because it is certainly easier to understand. Similarly, I refactored out the pivot_table calculation for `monetary_value` because that was also a bit hard to follow.